### PR TITLE
pageserver - add InMemoryLayer global map

### PR DESCRIPTION
**Pull request by patins**
A more naive approach, probably not useful for our current project, here - https://github.com/zenithdb/zenith/compare/layer-id

### CODE REVIEW

**Comment by hlinnaka** [comment]
For context to others:

We talked about how to solve the current acute problem in the page server: running out of memory. The plan is to:

1. Track open (i.e. in-memory) layers globally, across all tenants and timelines. Whenever an open in-memory layer is created, it's assigned a globally unique layer ID that can be used to reference it later. Currently we're passing around Arc<InMemoryLayer> references which is a bit awkward.

2. Cap the number of open layers. If there are more than X open layers, and another one is created, evict one of the old open layers to disk. The number of open layers is a crude proxy for the amount of memory used. The eviction is performed in the WAL receiver. Use a simple LRU or clock algorithm to choose the victim layer to evict. In reality, an open layer can consume anywhere between 1 kB and multiple GB of memory, with no hard upper bound. We should  track memory usage more accurately, but for starters, let's cap the number of layers.

3. Track memory usage more precisely.

4. Also address memory usage of on-disk layers. Each on-disk layer consumes some memory to track that it exists.

This PR does 1.

**Comment by hlinnaka** [comment]
Added a commit to add some comments.

**Comment by hlinnaka** [comment]
I've been working on steps 2-3 today, on top of this PR. Here: https://github.com/zenithdb/zenith/tree/layer-cache2

**Comment by hlinnaka** [comment]
Wrt. naming, I'd suggest:

`cache.rs` -> `global_layer_map.rs`

`struct Cache` -> `struct OpenLayers` or `struct OpenLayerRegistry`

`SlotId` -> `LayerId` or `OpenLayerId` (it identifies the layer in the slot, not the slot itself)

And maybe rename `version` to `tag`, I think that's the more commonly used term with the ABA problem.

File: pageserver/src/layered_repository/layer_map.rs [hunk 1]
```diff
@@ -77,7 +82,7 @@ impl LayerMap {
     pub fn insert_open(&mut self, layer: Arc<InMemoryLayer>) {
         let segentry = self.segs.entry(layer.get_seg_tag()).or_default();
 
-        segentry.update_open(Arc::clone(&layer));
+        let slot_id = segentry.update_open(Arc::clone(&layer));
```
Comment (hlinnaka) [inline pageserver/src/layered_repository/layer_map.rs:85]: `layer_id` here too, I think

**Review by hlinnaka — APPROVED** [review]
Let's get this in

File: pageserver/src/layered_repository/global_layer_map.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        slot.data = SlotData::Occupied(layer);
+
+        LayerId {
+            index: slot_idx,
+            tag: slot.tag,
+        }
+    }
+
+    pub fn get(&self, slot_id: &LayerId) -> Option<Arc<InMemoryLayer>> {
```
Comment (hlinnaka) [inline pageserver/src/layered_repository/global_layer_map.rs:79]: `slot_id` -> `layer_id` here too

Comment (hlinnaka) [reply pageserver/src/layered_repository/global_layer_map.rs:79]: and in remove()

Comment (patins) [reply pageserver/src/layered_repository/global_layer_map.rs:79]: whoops, will do

**Pull request merged**